qubesappmenus/__init__.py

```diff
@@ -210,6 +210,71 @@ def get_available(self, vm, fields=None, template=None):
                     result.append(field_values.get(field, ''))
             yield result
 
+    qubes_vm_desktop = 'org.qubes-os.qubes-vm'
+    qubes_vm_desktop_settings = 'org.qubes-os.qubes-vm-settings'
+
+    def desktop_name(self, vm, appmenu_basename: str):
+        """Return the basename of a ``.desktop`` file.
+
+        :param vm: QubesVM object for the VM whose entry this is.
+        :param appmenu_basename: The basename of the appmenu, as provided by
+        the VM.  This is not marked as “untrusted” as it has already been
+        validated.
+        """
+        assert appmenu_basename.endswith('.desktop')
+        return '.'.join((self.qubes_vm_desktop, str(vm), appmenu_basename))
+
+    def settings_name(self, vm):
+        """Return the basename of the .desktop file for the “Qube Settings”
+        menu entry of a given VM..
+
+        :param vm: QubesVM object for the VM.
+        """
+        return '.'.join((self.qubes_vm_desktop_settings, str(vm), 'desktop'))
+
+    def _do_remove_appmenus(self, vm, appmenus_to_remove, appmenus_dir,
+            refresh_cache):
+        """Remove old appmenus
+
+        :param vm: QubesVM object for the VM.
+        :param appmenus_to_remove: List of appmenus to remove.
+        :param appmenus_dir: Appmenus directory.
+        :param refresh_cache: Refresh the cacue?
+        """
+        old_appmenus = set((fname for fname in appmenus_to_remove
+                                  if self._is_old_path(fname)))
+        new_appmenus = appmenus_to_remove.difference(old_appmenus)
```

Comment:
`AttributeError: 'list' object has no attribute 'difference'`
```suggestion
        new_appmenus = set(appmenus_to_remove).difference(old_appmenus)
```
?

Reply:
Applied